### DISCUSSION TITLE
Why are Plans serialized explicitly in RetrieveRequest and SearchRequest (internal proto)

### DISCUSSION THREAD

**Discussion by ashkrisk** [milvus-io/milvus#35770] [Q&A and General discussion] [answered]
The [proto definitions](https://github.com/milvus-io/milvus/blob/56b3490a980a2ea3d03412e3c5d504bbc68eff89/internal/proto/internal.proto#L179) for `RetrieveRequest` and `SearchRequest` store the search/query plan in the following form:
```proto
// milvus/internal/proto/internal.proto
message RetrieveRequest {
  ...
  bytes serialized_expr_plan = 6;
  ...
}
```

The other alternative would be embed the `PlanNode` object directly in the `RetrieveRequest`, avoiding the need for explicit `Marshal`/`Unmarshal`:
```proto
message RetrieveRequest {
  ...
  plan.PlanNode plan = 6;
  ...
}
```

Why is the first version (explicit marshaling) preferred over the second?

**Comment by yhmo** [comment, marked as answer]
The reason is:
SearchRequest is created by proxy node, and passed to the shard leader node, finally passed to a worker query node. Only the worker query node decodes the `serialized_expr_plan`.
If we use the second approach, the 'plan' will be encoded/decoded twice.

**Reply by ashkrisk** [reply]
Okay, that makes sense. But is a `planpb.PlanNode` large enough to make add a significant encode/decode overhead?

For some context, I'm currently working on implementing https://github.com/milvus-io/milvus/issues/34754. As part of that, I'm working on an interface which can be used to add support for aggregation functions other than `count(*)`, such as `distinct(my_field)` as described in the issue, and potentially others like `max`, `min` etc.

`RetrieveResults` are reduced at multiple points throughout the query lifecycle, including at the shard leader. At that point, a different reducer needs to be chosen depending on which aggregation function (if any) is in use. The most straightforward way to do this is to decode the query plan and fetch the details from there.

The other option would be to have a extra field in the `RetrieveRequest` which contains all this information. This is similar to the how the `count(*)` operation can figure out it needs a different reducer: there is an `is_count` boolean field in `RetrieveRequest`. This avoids decoding the plan at the shard leader but leads to some information duplication which is not ideal.

@yhmo any thoughts? Also @congqixia

**Reply by yhmo** [reply]
Sometimes the filter expression might be complicated and the cost of encode/decode of PlanNode becomes expensive. If there are only a few vectors match the filter condition, the cost of encode/decode PlanNode might become a bottleneck.

In my opinion, the basic principle is: only encode/decode it when necessary.
The "is_count" of RetrieveRequest is a flag for count(*) to avoid encode/decode for the plan. I think it is a clever approach, but the proto structure should be carefully designed. If let me design the proto, I will prefer using a map<string, string> to store the flags like "is_count".

**Reply by ashkrisk** [reply]
Thanks @yhmo, I'll keep this in mind while implementing
